### PR TITLE
Fi 480.add task to setup permissions

### PR DESCRIPTION
**Pull request by Blackbaud-JasonBodnar**
@blackbaud/java-internal-tooling
@blackbaud/highlander

### CODE REVIEW

File: src/componentTest/groovy/com/blackbaud/templates/tasks/TemplateGenerationSpec.groovy [hunk 1]
```diff
@@ -434,5 +435,30 @@ class TemplateGenerationSpec extends AbstractProjectSpecification {
         greenwashOrAssertExpectedContent(restProject, "add-multiple-coreconfig-annotations")
     }
 
+    def "should thrown an exception if addPermissions task is called on a project that doesn't have a rest-client"() {
```
Comment (Blackbaud-BethRichardson) [inline src/componentTest/groovy/com/blackbaud/templates/tasks/TemplateGenerationSpec.groovy]: typo in test name "thrown"

File: src/main/resources/templates/springboot/rest/permissions/service-permissions.java.tmpl [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+import com.blackbaud.security.permission.AbstractPermissions;
+
+import java.util.Map;
+
+public class ${serviceName}Permissions extends AbstractPermissions {
+    @Override
+    protected void initPermissions(Map<String, Long> permissions) {
+        permissions.put("${servicePermissionPrefix}.view", 1L);
+        permissions.put("${servicePermissionPrefix}.edit", 2L);
+        permissions.put("${servicePermissionPrefix}.delete", 3L);
```
Comment (Blackbaud-BethRichardson) [inline src/main/resources/templates/springboot/rest/permissions/service-permissions.java.tmpl]: Should we have a comment that these need to be changed?

File: src/main/groovy/com/blackbaud/templates/project/BasicProject.groovy [hunk 3]
```diff
@@ -321,4 +319,4 @@ lombok.addLombokGeneratedAnnotation = true
 """
     }
 
-}
```
Comment (Blackbaud-BethRichardson) [inline src/main/groovy/com/blackbaud/templates/project/BasicProject.groovy:324]: is this change needed?

**Review by Blackbaud-BethRichardson — CHANGES_REQUESTED** [review]

File: src/main/resources/templates/springboot/rest/permissions/service-permissions.java.tmpl (hunk 2 above)
Comment (Blackbaud-MikeLueders) [reply src/main/resources/templates/springboot/rest/permissions/service-permissions.java.tmpl]: consider not having dummy permissions - if it's unclear what should go here, add a comment describing the intended content of the method

File: src/main/resources/templates/springboot/rest/permissions/permissions-config.java.tmpl [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+@Configuration
+public class PermissionsConfig {
+    @Autowired
+    private PermissionsRegistry permissionsRegistry;
+
+    @PostConstruct
+    private void registerPermissions() {
+        permissionsRegistry.registerPermissionConfiguration(
+            new ${serviceName}Permissions()
```
Comment (Blackbaud-MikeLueders) [inline src/main/resources/templates/springboot/rest/permissions/permissions-config.java.tmpl]: this would generally be on a single line

File: src/componentTest/groovy/com/blackbaud/templates/tasks/TemplateGenerationSpec.groovy (hunk 1 above)
Comment (Blackbaud-JasonBodnar) [reply src/componentTest/groovy/com/blackbaud/templates/tasks/TemplateGenerationSpec.groovy]: Fixed.

Comment (Blackbaud-JasonBodnar) [reply src/componentTest/groovy/com/blackbaud/templates/tasks/TemplateGenerationSpec.groovy]: Fixed.

File: src/main/groovy/com/blackbaud/templates/project/PermissionsProject.groovy [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        basicProject.applyTemplate("src/main/java/${servicePackagePath}/core") {
+            "PermissionsConfig.java" template: "/templates/springboot/rest/permissions/permissions-config.java.tmpl",
+                                packageName: servicePackage,
+                                serviceName: serviceName
+        }
+    }
+
+    private void addGlobalMethodSecurityConfigToCoreConfig() {
+        ProjectFile coreConfigClass = basicProject.findFile("CoreConfig.java")
+        coreConfigClass.addConfigurationImport("com.blackbaud.security.GlobalMethodSecurityConfig")
```
Comment (Blackbaud-MikeLueders) [inline src/main/groovy/com/blackbaud/templates/project/PermissionsProject.groovy]: isn't this imported via common-deployable?

File: src/main/resources/templates/springboot/rest/permissions/service-permissions.java.tmpl (hunk 2 above)
Comment (Blackbaud-JasonBodnar) [reply src/main/resources/templates/springboot/rest/permissions/service-permissions.java.tmpl]: Added.

File: src/main/groovy/com/blackbaud/templates/project/BasicProject.groovy (hunk 3 above)
Comment (Blackbaud-JasonBodnar) [reply src/main/groovy/com/blackbaud/templates/project/BasicProject.groovy:324]: Yes,

File: src/main/groovy/com/blackbaud/templates/project/PermissionsProject.groovy [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    private void addPermissionsConfigToCoreConfig() {
+        ProjectFile coreConfigClass = basicProject.findFile("CoreConfig.java")
+        coreConfigClass.addConfigurationImport("${servicePackage}.core.PermissionsConfig", true)
```
Comment (Blackbaud-MikeLueders) [inline src/main/groovy/com/blackbaud/templates/project/PermissionsProject.groovy]: permissions are not a core concern, should go on main class not CoreConfig

File: src/main/groovy/com/blackbaud/templates/project/PermissionsProject.groovy [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+    private String getServicePackage() {
+        basicProject.servicePackage
+    }
+
+    private String getServicePackagePath() {
+        basicProject.servicePackagePath
+    }
+
+    void addPermissions() {
```
Comment (Blackbaud-MikeLueders) [inline src/main/groovy/com/blackbaud/templates/project/PermissionsProject.groovy]: permissions are solely a rest project concern, consider moving these methods to `RestProject`

File: src/main/resources/templates/springboot/rest/permissions/service-permissions.java.tmpl (hunk 2 above)
Comment (Blackbaud-BethRichardson) [reply src/main/resources/templates/springboot/rest/permissions/service-permissions.java.tmpl]: Could you add a "TODO:" so that it stands out in intellij that this should be changed?

**Review by Blackbaud-MikeLueders — CHANGES_REQUESTED** [review]

**Review by Blackbaud-BethRichardson — APPROVED** [review]

File: src/main/resources/templates/springboot/rest/permissions/permissions-config.java.tmpl (hunk 4 above)
Comment (Blackbaud-JasonBodnar) [reply src/main/resources/templates/springboot/rest/permissions/permissions-config.java.tmpl]: Fixed.

File: src/main/groovy/com/blackbaud/templates/project/PermissionsProject.groovy (hunk 5 above)
Comment (Blackbaud-JasonBodnar) [reply src/main/groovy/com/blackbaud/templates/project/PermissionsProject.groovy]: Yes. It no longer needs to be needed now that `PermissionsConfig` is added to the main class.

File: src/main/groovy/com/blackbaud/templates/project/PermissionsProject.groovy (hunk 6 above)
Comment (Blackbaud-JasonBodnar) [reply src/main/groovy/com/blackbaud/templates/project/PermissionsProject.groovy]: Fixed.

File: src/main/groovy/com/blackbaud/templates/project/PermissionsProject.groovy (hunk 7 above)
Comment (Blackbaud-JasonBodnar) [reply src/main/groovy/com/blackbaud/templates/project/PermissionsProject.groovy]: Moved.

**Review by Blackbaud-MikeLueders — APPROVED** [review]
